doctor/lib/text_extraction.py

```diff
@@ -0,0 +1,326 @@
+            page_text = (
+                page_text[:location]
+                + " " * (longest - index)
+                + page_text[location:]
+            )
+            adjust += longest - index
+        return page_text
+    return page_text
+
+
+def page_needs_ocr(page: pdfplumber.pdf.Page, page_text: str) -> bool:
+    """Does the page need OCR
+
+    :param page:Pdf Plumber Page
+    :param page_text: context extracted from page
+    :return: does page need OCR
+    """
+    if (
+        page_text.strip() == ""
+        or "(cid:" in page_text
+        or page_annotations(page)
+        or page_images(page)
+        or len(page.curves) > 10
+    ):
+        return True
+    return False
+
+
+def convert_pdf_page_to_image(
+    page: pdfplumber.pdf.Page, strip_margin: bool
+) -> Image:
+    """Conver page to image and crop margin if applicable
+
+    :param page:Pdf Plumber
+    :param strip_margin: bool
+    :return: The formatted image
+    """
+    img = page.to_image(resolution=300)
+    _, _, w, h = page.bbox
+    width = w * img.scale
+    height = h * img.scale
+
+    if strip_margin == True:
+        # Because this is OCR - I think its reasonable to crop half the standard
+        # 1 inch margin - this leads to much better results and helps reduce
+        bbox = (
+            (1 / (8.5 * 2) * width),
+            (1 / (11 * 2) * height),
+            (16 / (8.5 * 2) * width),
+            (21 / (11 * 2) * height),
+        )
+        image = img.original.crop(bbox)
+    else:
+        image = img.original
+    return image
+
+
+def ocr_image_to_data(image: Image) -> list[pd.DataFrame]:
+    """Perform OCR on an image to extract data
+
+    Convert the image of the pdf page to OCR data
+    :param image: Pil Image
+    :return: A list of DataFrames, each containing OCR data for a block of text
+    """
+    data_dict = pytesseract.image_to_data(
+        image,
+        config="-c preserve_interword_spaces=1x1 -c tessedit_do_invert=0 --psm 6 -l eng",
+        output_type=Output.DICT,
+    )
+    df = pd.DataFrame(data_dict)
+    filtered_data = df[(df.conf != -1)]
+    block_ids = (
+        filtered_data.groupby("block_num")
+        .first()
+        .sort_values("top")
+        .index.tolist()
+    )
+    blocks = [
+        filtered_data[filtered_data["block_num"] == block]
+        for block in block_ids
+    ]
+    return blocks
+
+
+def extract_with_ocr(page: pdfplumber.pdf.Page, strip_margin: bool) -> str:
+    """Extract the page using OCR
+
+    :param page:Pdf Plumber Page
+    :param strip_margin: If we should trim the margins
+    :return: The extracted content for the page
+    """
+
+    image = convert_pdf_page_to_image(page, strip_margin)
+    data = ocr_image_to_data(image)
+    content = ""
+    prev = {}
+    for words in data:
+        for index, word in words.iterrows():
+            content = insert_whitespace(content, word, prev)
+            content += get_word(word, image.size[0], strip_margin)
```

Comment:
We should make some unit tests for insert_whitespace and get_word, because they're finicky, an easy place for bugs, and a difficult thing to work with in the future if you ever have to tweak them. A perfect place for TDD.